// @webacad/ng-mat-file-upload/esm5/upload-file.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { readFileAsDataURL } from '@webacad/observable-file-reader';
/** @type {?} */
var imageMimeTypes = [
    'image/gif', 'image/png', 'image/jpeg',
];
var UploadFile = /** @class */ (function () {
    function UploadFile(file, _progress) {
        if (_progress === void 0) { _progress = 0; }
        this.file = file;
        this._progress = _progress;
        this.icon = 'cloud_upload';
    }
    Object.defineProperty(UploadFile.prototype, "progress", {
        get: /**
         * @return {?}
         */
        function () {
            return this._progress;
        },
        set: /**
         * @param {?} progress
         * @return {?}
         */
        function (progress) {
            this._progress = progress;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    UploadFile.prototype.supportsImagePreview = /**
     * @return {?}
     */
    function () {
        return imageMimeTypes.indexOf(this.file.type) >= 0;
    };
    /**
     * @return {?}
     */
    UploadFile.prototype.getImageSource = /**
     * @return {?}
     */
    function () {
        if (typeof this.imageSource !== 'undefined') {
            return this.imageSource;
        }
        return this.imageSource = readFileAsDataURL(this.file);
    };
    /**
     * @param {?} add
     * @return {?}
     */
    UploadFile.prototype.increaseProgress = /**
     * @param {?} add
     * @return {?}
     */
    function (add) {
        this.progress += add;
    };
    return UploadFile;
}());
export { UploadFile };
if (false) {
    /** @type {?} */
    UploadFile.prototype.icon;
    /** @type {?} */
    UploadFile.prototype.imageSource;
    /** @type {?} */
    UploadFile.prototype.file;
    /** @type {?} */
    UploadFile.prototype._progress;
}
