// @webacad/ng-mat-file-upload/esm5/validators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @param {?} maxSize
 * @return {?}
 */
export function fileMaxSize(maxSize) {
    return fileValidation(function (file) { return file.size > maxSize; }, function (errors) {
        return {
            fileMaxSize: {
                maxSize: maxSize,
                invalidFiles: errors,
            },
        };
    });
}
/**
 * @param {?} allowedTypes
 * @return {?}
 */
export function fileType(allowedTypes) {
    return fileValidation(function (file) { return allowedTypes.indexOf(file.type) < 0; }, function (errors) {
        return {
            fileType: {
                allowedTypes: allowedTypes,
                invalidFiles: errors,
            },
        };
    });
}
/**
 * @param {?} validator
 * @param {?} createErrors
 * @return {?}
 */
function fileValidation(validator, createErrors) {
    return function (control) {
        if (!control.value) {
            return null;
        }
        /** @type {?} */
        var files = control.value;
        /** @type {?} */
        var errors = [];
        for (var i = 0; i < files.length; i++) {
            if (validator(files[i])) {
                errors.push(files[i]);
            }
        }
        if (errors.length) {
            return createErrors(errors);
        }
        return null;
    };
}
