// @webacad/ng-mat-file-upload/esm2015/upload-file.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { readFileAsDataURL } from '@webacad/observable-file-reader';
/** @type {?} */
const imageMimeTypes = [
    'image/gif', 'image/png', 'image/jpeg',
];
export class UploadFile {
    /**
     * @param {?} file
     * @param {?=} _progress
     */
    constructor(file, _progress = 0) {
        this.file = file;
        this._progress = _progress;
        this.icon = 'cloud_upload';
    }
    /**
     * @return {?}
     */
    get progress() {
        return this._progress;
    }
    /**
     * @param {?} progress
     * @return {?}
     */
    set progress(progress) {
        this._progress = progress;
    }
    /**
     * @return {?}
     */
    supportsImagePreview() {
        return imageMimeTypes.indexOf(this.file.type) >= 0;
    }
    /**
     * @return {?}
     */
    getImageSource() {
        if (typeof this.imageSource !== 'undefined') {
            return this.imageSource;
        }
        return this.imageSource = readFileAsDataURL(this.file);
    }
    /**
     * @param {?} add
     * @return {?}
     */
    increaseProgress(add) {
        this.progress += add;
    }
}
if (false) {
    /** @type {?} */
    UploadFile.prototype.icon;
    /** @type {?} */
    UploadFile.prototype.imageSource;
    /** @type {?} */
    UploadFile.prototype.file;
    /** @type {?} */
    UploadFile.prototype._progress;
}
