// @webacad/ng-mat-file-upload/esm2015/validators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @param {?} maxSize
 * @return {?}
 */
export function fileMaxSize(maxSize) {
    return fileValidation((file) => file.size > maxSize, (errors) => {
        return {
            fileMaxSize: {
                maxSize: maxSize,
                invalidFiles: errors,
            },
        };
    });
}
/**
 * @param {?} allowedTypes
 * @return {?}
 */
export function fileType(allowedTypes) {
    return fileValidation((file) => allowedTypes.indexOf(file.type) < 0, (errors) => {
        return {
            fileType: {
                allowedTypes: allowedTypes,
                invalidFiles: errors,
            },
        };
    });
}
/**
 * @param {?} validator
 * @param {?} createErrors
 * @return {?}
 */
function fileValidation(validator, createErrors) {
    return (control) => {
        if (!control.value) {
            return null;
        }
        /** @type {?} */
        const files = control.value;
        /** @type {?} */
        const errors = [];
        for (let i = 0; i < files.length; i++) {
            if (validator(files[i])) {
                errors.push(files[i]);
            }
        }
        if (errors.length) {
            return createErrors(errors);
        }
        return null;
    };
}
