// angular-google-charts/esm5/lib/helpers/google-chart-packages.helper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var GoogleChartPackagesHelper = /** @class */ (function () {
    function GoogleChartPackagesHelper() {
    }
    /**
     * @param {?} chartName
     * @return {?}
     */
    GoogleChartPackagesHelper.getPackageForChartName = /**
     * @param {?} chartName
     * @return {?}
     */
    function (chartName) {
        return GoogleChartPackagesHelper.ChartTypesToPackages[chartName];
    };
    GoogleChartPackagesHelper.ChartTypesToPackages = {
        AnnotationChart: 'annotationchart',
        AreaChart: 'corechart',
        Bar: 'bar',
        BarChart: 'corechart',
        BubbleChart: 'corechart',
        Calendar: 'calendar',
        CandlestickChart: 'corechart',
        ColumnChart: 'corechart',
        ComboChart: 'corechart',
        PieChart: 'corechart',
        Gantt: 'gantt',
        Gauge: 'gauge',
        GeoChart: 'geochart',
        Histogram: 'corechart',
        Line: 'line',
        LineChart: 'corechart',
        Map: 'map',
        OrgChart: 'orgchart',
        Sankey: 'sankey',
        Scatter: 'scatter',
        ScatterChart: 'corechart',
        SteppedAreaChart: 'corechart',
        Table: 'table',
        Timeline: 'timeline',
        TreeMap: 'treemap',
        WordTree: 'wordtree'
    };
    return GoogleChartPackagesHelper;
}());
export { GoogleChartPackagesHelper };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GoogleChartPackagesHelper.ChartTypesToPackages;
}
